File: app/src/main/java/com/example/sy43/db/DAO/TransactionDAO.java
package com.example.sy43.db.DAO;


import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.example.sy43.db.entity.Transaction;
import com.google.common.util.concurrent.ListenableFuture;

import java.util.List;
import androidx.room.Dao;

@Dao
public interface TransactionDAO{
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public ListenableFuture<Void> insert(Transaction... trans);

    @Query("SELECT * FROM `transaction` WHERE TransactionID LIKE :transactionID")
    public ListenableFuture<List<com.example.sy43.db.entity.Transaction>> findByID(int transactionID);

    @Query("SELECT Value FROM `transaction`")
    public ListenableFuture<List<Float>> findValues();

    @Query("SELECT * FROM `transaction` WHERE Category LIKE :category")
    public ListenableFuture<List<com.example.sy43.db.entity.Transaction>> findByCategory(int category);

    @Query("SELECT * FROM `transaction` WHERE SubCategory LIKE :subcategory")
    public ListenableFuture<List<com.example.sy43.db.entity.Transaction>> findBySubCategory(int subcategory);

    @Delete
    void deleteTransaction(com.example.sy43.db.entity.Transaction transaction);
}

File: Fluz/app/src/main/java/com/example/fluz/data/dao/TransactionDao.kt
package com.example.fluz.data.dao

import androidx.room.*
import com.example.fluz.data.entities.Transaction
import com.example.fluz.data.relashionships.TransactionAndCategory
import kotlinx.coroutines.flow.Flow

/**
 * SQL Requests for Transaction entity
 */
@Dao
interface TransactionDao {

    /**
     * Insert a new transaction
     *
     * @param transaction the transaction to insert
     */
    @Insert(onConflict = OnConflictStrategy.ABORT)
    suspend fun insert(transaction: Transaction)

    /**
     * Select one transaction with its category
     *
     * @param transactionId the id of the transaction
     * @return the select transaction with its category
     */
    @androidx.room.Transaction
    @Query("SELECT * FROM 'Transaction' WHERE id =:transactionId")
    fun getWithCategory(transactionId: Int): Flow<TransactionAndCategory>

    /**
     * Select all transaction
     *
     * @return a list of all the transactions
     */
    @Query("SELECT * FROM 'Transaction'")
    fun getAll(): Flow<List<Transaction>>

    /**
     * Delete one transaction with transaction id
     *
     * @param transactionId the id of the transaction
     */
    @Query("DELETE FROM 'Transaction' WHERE id = :transactionId")
    suspend fun delete(transactionId: Int)

    /**
     * Delete all categories
     *
     */
    @Query("DELETE FROM 'Transaction'")
    suspend fun deleteAll()
}